find/finder.go

```diff
@@ -113,9 +113,68 @@ func (f *Finder) findRoot(ctx context.Context, root *list.Element, parts []strin
 	return true
 }
 
+var refTypeMap = map[string]string{
+	"datacenter":  "Datacenter",
+	"datastore":   "Datastore",
+	"domain":      "ClusterComputeResource",
+	"dvportgroup": "DistributedVirtualPortgroup",
+	"dvs":         "DistributedVirtualSwitch",
+	"group":       "Folder",
+	"host":        "HostSystem",
+	"network":     "Network",
+	"resgroup":    "ResourcePool",
+	"vm":          "VirtualMachine",
+}
+
+var prefixTypeMap = map[string]struct{ prefix, kind string }{
+	"domain":   {"s", "ComputeResource"},
+	"group":    {"p", "StoragePod"},
+	"resgroup": {"v", "VirtualApp"},
+}
+
+func referenceFromString(arg string) *types.ManagedObjectReference {
+	if ref := object.ReferenceFromString(arg); ref != nil {
+		return ref
+	}
+
+	id := strings.SplitN(arg, "-", 2)
+	if len(id) != 2 {
+		return nil
+	}
+
+	if kind, ok := refTypeMap[id[0]]; ok {
```

Comment:
Just for my understanding, this is invalid when someone passes as an argument something like:

"datacenter-3" right?

So which specific case we fall into this? I am not down deep on vSphere API, I remember this may be a supervisor domain, just trying to understand what "argument" here is valid.

Reply:
No, "datacenter-3" is valid. 

`if object.ReferenceFromString` is true when value is a managed object reference (MOR). That is, it includes the `Type:` `Datacenter:datacenter-3`

This section parses managed object ID (MOID). That is, without the `Type:` - the type is derived from the ID value.

Reply:
Moved all this into `object.ReferenceFromString` as it isn't Finder specific, and included these comments there. `TestReferenceFromString` should also make this more clear.